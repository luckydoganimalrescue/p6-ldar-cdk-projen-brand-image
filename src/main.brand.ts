import path from "path";
import { Readable } from "stream";
import {
  GetObjectCommand,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import {
  SESClient,
  SendEmailCommand,
  SendEmailRequest,
} from "@aws-sdk/client-ses";
import AdmZip from "adm-zip";
import {
  APIGatewayProxyEvent,
  APIGatewayProxyResult,
  Context,
} from "aws-lambda";
import * as mime from "mime-types";
import sharp from "sharp";

// Constants
const compositeImageKeyWhite = "processed_LDARBranding-white.png";
const compositeImageKeyBlack = "processed_LDARBranding-black.png";
const width = 1400;
const height = 1400;

// Initialize AWS services
const s3 = new S3Client({});
const ses = new SESClient({});

// Lambda handler function
export const handler = async (
  event: APIGatewayProxyEvent,
  _context: Context
): Promise<APIGatewayProxyResult> => {
  try {
    const { results, email, zipKey } = await processRequest(event);
    await sendEmail(results, email, zipKey);
    const htmlContent = generateEmailHtml(results, zipKey);
    return generateResponse(200, htmlContent);
  } catch (error: any) {
    console.error(error);
    return generateResponse(
      500,
      `<h1>Error processing request</h1><p>${error.message}</p>`
    );
  }
};

const sendEmail = async (results: any[], email: string, zipKey: string) => {
  const htmlContent = generateEmailHtml(results, zipKey);
  const params: SendEmailRequest = {
    Source: process.env.EMAIL_SENDER || "your-sender-email@example.com",
    Destination: {
      ToAddresses: [email],
    },
    Message: {
      Subject: {
        Data: "P6 LDAR Pet Image Branding Results: Success",
      },
      Body: {
        Html: {
          Data: htmlContent,
        },
      },
    },
  };

  try {
    const command = new SendEmailCommand(params);
    await ses.send(command);
    console.log("Email sent successfully");
  } catch (error) {
    console.error("Failed to send email:", error);
  }
};

const generateEmailHtml = (results: any[], zipKey: string): string => {
  let html = "<html><body>";

  if (zipKey !== "") {
    const zipUrl = generateS3Url(
      process.env.BRAND_IMAGE_BUCKET ?? "p6-dne",
      zipKey
    );
    html += '<div style="margin-top: 20px;">';
    html += "<h3>Download Zip File</h3>";
    html += `<a href="${zipUrl}">${zipUrl}</a><br/>`;
    html += "</div><br/>";
  }

  for (const result of results) {
    html += '<div style="display: flex;">';
    html += '<div style="flex: 1;">';
    html += "<h3>Original Image</h3>";
    html += `<a href="${result.originalUrl}">${result.originalUrl}</a><br/>`;
    html += `<a href="${result.originalUrl}">`;
    html += `<img src="${result.originalUrl}" alt="Original Image" style="max-width: 400px;"/><br/>`;
    html += "</a>";
    html += "</div>";

    html += '<div style="flex: 1;">';
    html += "<h3>Processed Image</h3>";
    html += `<a href="${result.processedUrl}">${result.processedUrl}</a><br/>`;
    html += `<a href="${result.processedUrl}">`;
    html += `<img src="${result.processedUrl}" alt="Processed Image" style="max-width: 400px;"/><br/>`;
    html += "</a>";
    html += "</div>";
    html += "</div>";
  }

  html += "</body></html>";

  return html;
};

// Process the incoming request
const processRequest = async (
  event: APIGatewayProxyEvent
): Promise<{ results: any[]; email: string; zipKey: string }> => {
  const body = JSON.parse(event.body || "{}");
  const filename = body.image;
  const email = body.email;

  const buffer = await downloadFromS3(filename);
  const files = await extractFiles(buffer, filename);

  console.log("Processing files");
  const results = await Promise.all(files.map(processFile));
  console.log("Finished processing files");

  if (filename.endsWith(".zip")) {
    const zipBuffer = await zipFiles(results);
    const zipKey = generateKey("processed_files.zip", "package");
    await uploadToS3(
      process.env.BRAND_IMAGE_BUCKET ?? "p6-dne",
      zipKey,
      zipBuffer
    );
    return {
      results,
      email,
      zipKey,
    };
  } else {
    return { results, email, zipKey: "" };
  }
};

// Zip the processed files
const zipFiles = async (results: any[]): Promise<Buffer> => {
  console.log("Zipping files");
  const zip = new AdmZip();

  for (const result of results) {
    const bucketKey = result.processedUrl.substring(
      result.processedUrl.indexOf(".com/") + 5
    );
    const filename = bucketKey.split("/").pop();
    const fileContent = await downloadFromS3(bucketKey);
    console.log(
      `Zipping ${bucketKey} as ${filename} from ${result.processedUrl}`
    );
    zip.addFile(filename, fileContent);
  }

  return zip.toBuffer();
};

// Extract files from the request
const extractFiles = async (
  contents: Buffer,
  filename: string
): Promise<{ content: Buffer; filename: string }[]> => {
  const allowedFileTypes = [
    "application/zip",
    "image/gif",
    "image/jpeg",
    "image/png",
  ];
  const fileType = mime.lookup(filename);
  if (fileType !== false) {
    if (!allowedFileTypes.includes(fileType)) {
      throw new Error("Unsupported file type");
    }
  }

  if (filename.endsWith(".zip")) {
    return extractFilesFromZip(contents);
  } else {
    return [{ content: contents, filename: filename }];
  }
};

// Extract files from a zip archive
const extractFilesFromZip = async (zipFileContent: Buffer): Promise<any> => {
  const zip = new AdmZip(zipFileContent);
  const entries = zip.getEntries();
  const extractedFiles = [];

  for (const entry of entries) {
    if (entry.isDirectory || entry.entryName.includes("__MACOSX")) continue;
    const extension = path.extname(entry.entryName).toLowerCase();
    if (![".jpg", ".gif", ".png"].includes(extension)) continue;

    const filename = entry.entryName;
    const content = entry.getData();

    const extractedFile = {
      content,
      filename,
      contentType: "",
      encoding: "",
      fieldname: "",
    };

    extractedFiles.push(extractedFile);
  }

  return extractedFiles;
};

// Process a single file
const processFile = async (file: {
  content: Buffer;
  filename: string;
}): Promise<any> => {
  console.log("Processing file:", file.filename);
  const processedFile = await processImage(file.content);

  const bucket = process.env.BRAND_IMAGE_BUCKET ?? "p6-dne";
  const originalKey = generateKey(file.filename, "original");
  const processedKey = generateKey(file.filename, "processed");
  await uploadToS3(bucket, originalKey, file.content);
  await uploadToS3(bucket, processedKey, processedFile);

  const originalUrl = generateS3Url(bucket, originalKey);
  const processedUrl = generateS3Url(bucket, processedKey);

  return {
    originalUrl,
    processedUrl,
  };
};

// Process the image (resize and watermark)
const processImage = async (content: Buffer): Promise<Buffer> => {
  console.log("Processing Image...");
  const image = sharp(content);
  const resizedBuffer = await resizeImage(image);
  const processedBuffer = await addWatermark(resizedBuffer);

  return processedBuffer;
};

// Resize the image to the desired dimensions
const resizeImage = async (image: sharp.Sharp): Promise<Buffer> => {
  console.log("Resizing Image....");
  return image
    .resize({
      fit: sharp.fit.inside,
      width,
      height,
    })
    .toBuffer();
};

// A utility function to handle caching
const watermarkCache: { [key: string]: Buffer } = {}; // Define the cache object
const getWatermarkBuffer = async (key: string): Promise<Buffer> => {
  if (watermarkCache[key]) {
    // If the watermark is in cache, use that
    return watermarkCache[key];
  } else {
    // If not, download it, put it in the cache, and return it
    const buffer = await downloadFromS3(key);
    watermarkCache[key] = buffer;
    return buffer;
  }
};

// Add the watermark to the image
const addWatermark = async (imageBuffer: Buffer): Promise<Buffer> => {
  console.log("Adding WaterMark...");
  const watermarkKey = await getWatermarkKey(imageBuffer);
  const watermarkBuffer = await getWatermarkBuffer(watermarkKey);
  const { width: imageWidth, height: imageHeight } = await sharp(
    imageBuffer
  ).metadata();
  const { width: watermarkWidth, height: watermarkHeight } = await sharp(
    watermarkBuffer
  ).metadata();

  // Position adjustment based on watermark color
  let left = (imageWidth ?? 0) - (watermarkWidth ?? 0) + 25;
  let top = (imageHeight ?? 0) - (watermarkHeight ?? 0) + 45;

  // Adjust position if the watermark is white
  if (watermarkKey === compositeImageKeyWhite) {
    left -= 30;
    top -= 40;
  }

  // Adjust position if the watermark is black
  if (watermarkKey === compositeImageKeyBlack) {
    top -= 50;
  }

  console.log(`Watermark position: ${left}, ${top}`);
  return sharp(imageBuffer)
    .composite([{ input: watermarkBuffer, left, top }])
    .toBuffer();
};

// Get the watermark key based on image luminance
const getWatermarkKey = async (imageBuffer: Buffer): Promise<string> => {
  const { dominant } = await sharp(imageBuffer).stats();
  const { r, g, b } = dominant;
  const luminance = (r * 0.299 + g * 0.587 + b * 0.114) / 255;
  const key = luminance < 0.5 ? compositeImageKeyWhite : compositeImageKeyBlack;
  console.log(`Watermark key: ${key}`);
  return key;
};

// Generate an S3 URL for the given bucket and key
const generateS3Url = (bucket: string, key: string): string => {
  const url = `https://${bucket}.s3.amazonaws.com/${key}`;
  console.log(`Generated S3 URL: ${url}`);
  return url;
};

// Upload a file to S3
const uploadToS3 = async (
  bucket: string,
  key: string,
  content: Buffer
): Promise<void> => {
  console.log(`Uploading ${bucket}/${key}`);
  const command = new PutObjectCommand({
    Bucket: bucket,
    Key: key,
    Body: content,
  });
  await s3.send(command);
};

// Download a file from S3
const downloadFromS3 = async (key: string): Promise<Buffer> => {
  const bucket = process.env.BRAND_IMAGE_BUCKET ?? "p6-dne";
  const params = {
    Bucket: bucket,
    Key: key,
  };
  console.log(`Downloading ${bucket}/${key}`);
  const command = new GetObjectCommand(params);
  let response = await s3.send(command);
  if (!response.Body) {
    console.error(
      `Failed to download file. Response: ${JSON.stringify(response, null, 2)}`
    );
    throw new Error(`Failed to download file from S3: ${bucket}/${key}`);
  } else {
    console.log(`Successfully downloaded file from S3: ${bucket}/${key}`);
    return streamToBuffer(response.Body as Readable);
  }
};

// Convert a stream to a buffer
const streamToBuffer = (stream: Readable): Promise<Buffer> => {
  return new Promise((resolve, reject) => {
    const chunks: any[] = [];
    stream.on("data", (chunk) => chunks.push(chunk));
    stream.on("error", reject);
    stream.on("end", () => resolve(Buffer.concat(chunks)));
  });
};

// Generate a key for an image (original or processed)
const generateKey = (
  filename: string,
  type: "original" | "processed" | "package"
): string => {
  const currentDateTime = new Date().toISOString().replace(/:/g, "-");
  const [currentDate, currentTime] = currentDateTime.split("T");
  const formattedDateTime = `${currentDate}_${currentTime.split(".")[0]}_${
    currentTime.split(".")[1]
  }`;
  const prefix = type;
  const key = `${formattedDateTime}_${prefix}_${filename}`;
  console.log(`Generated key: ${key}`);
  return key;
};

// Generate a response with the given status code and body
const generateResponse = (
  statusCode: number,
  body: string
): APIGatewayProxyResult => {
  return {
    statusCode,
    body,
    headers: {
      "Content-Type": "text/html",
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Headers": "*",
      "Access-Control-Allow-Methods": "GET,PUT,POST,DELETE,PATCH,OPTIONS",
    },
  };
};
