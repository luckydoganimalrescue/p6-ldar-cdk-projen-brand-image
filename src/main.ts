import * as apigw from "@aws-cdk/aws-apigatewayv2-alpha";
import * as apigwi from "@aws-cdk/aws-apigatewayv2-integrations-alpha";
import * as cdk from "aws-cdk-lib";
import * as certificatemanager from "aws-cdk-lib/aws-certificatemanager";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as lambdajs from "aws-cdk-lib/aws-lambda-nodejs";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as ses from "aws-cdk-lib/aws-ses";
import * as floyd from "cdk-iam-floyd";
import { Construct } from "constructs";

const HOSTED_ZONE_NAME = "p6m7g8.net";
const VERIFY_EMAIL = `pgollucci@${HOSTED_ZONE_NAME}`;
const SUBDOMAIN_NAME = "ldar";
const FROM_EMAIL = "ldar-pet-brander@p6m7g8.com";
const RECORD_NAME = SUBDOMAIN_NAME + "." + HOSTED_ZONE_NAME;

export class MyStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: cdk.StackProps = {}) {
    super(scope, id, props);

    // const hostedZone = route53.HostedZone.fromLookup(this, "HostedZone", {
    //   domainName: HOSTED_ZONE_NAME,
    // });

    const certificate = new certificatemanager.Certificate(
      this,
      "Certificate",
      {
        domainName: RECORD_NAME,
        validation: certificatemanager.CertificateValidation.fromEmail({
          email: VERIFY_EMAIL,
        }),
      }
    );

    const domainName = new apigw.DomainName(this, "DN", {
      domainName: RECORD_NAME,
      certificate: certificate,
    });

    const senderEmail = ses.Identity.email(FROM_EMAIL);
    new ses.EmailIdentity(this, "Identity", {
      identity: senderEmail,
    });

    const bucket = new s3.Bucket(this, "MyBucket", {
      accessControl: s3.BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ACLS,
      publicReadAccess: true,
      websiteIndexDocument: "index.html",
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      cors: [
        {
          allowedMethods: [s3.HttpMethods.GET, s3.HttpMethods.PUT],
          allowedOrigins: ["*"],
          allowedHeaders: ["*"],
          maxAge: 3000,
        },
      ],
    });

    const presignFunc = new lambdajs.NodejsFunction(this, "presign", {
      runtime: lambda.Runtime.NODEJS_18_X,
    });
    bucket.grantPut(presignFunc);
    bucket.grantPublicAccess("*", "s3:PutObject");

    presignFunc.addEnvironment("BUCKET_NAME", bucket.bucketName);

    const brandFunc = new lambdajs.NodejsFunction(this, "brand", {
      runtime: lambda.Runtime.NODEJS_18_X,
      tracing: lambda.Tracing.ACTIVE,
      timeout: cdk.Duration.minutes(14),
      memorySize: 512,
      bundling: {
        nodeModules: ["sharp"],
        forceDockerBundling: true,
        //        minify: true,
      },
    });
    bucket.grantReadWrite(brandFunc);

    brandFunc.addEnvironment("BRAND_IMAGE_BUCKET", bucket.bucketName);
    brandFunc.addEnvironment("EMAIL_SENDER", FROM_EMAIL);
    brandFunc.addEnvironment("EMAIL_REGION", "us-east-1");
    brandFunc.addEnvironment("FIT", "fill");

    const policy = new floyd.Iam().allow().to("ses:SendEmail").on("*");
    brandFunc.addToRolePolicy(policy);

    const httpApi = new apigw.HttpApi(this, "HttpApi", {
      description: "Pet Brander API",
      corsPreflight: {
        allowHeaders: ["*"],
        allowMethods: [apigw.CorsHttpMethod.ANY],
        allowOrigins: ["*"],
        maxAge: cdk.Duration.days(10),
      },
      defaultDomainMapping: {
        domainName: domainName,
      },
    });

    const presignIntegration = new apigwi.HttpLambdaIntegration(
      "FuncIntegration",
      presignFunc
    );
    httpApi.addRoutes({
      path: "/presign",
      methods: [apigw.HttpMethod.POST],
      integration: presignIntegration,
    });

    const brandIntegration = new apigwi.HttpLambdaIntegration(
      "brandIntegration",
      brandFunc
    );
    httpApi.addRoutes({
      path: "/brand",
      methods: [apigw.HttpMethod.POST],
      integration: brandIntegration,
    });
  }
}

const theEnv = {
  account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
};

const app = new cdk.App();
new MyStack(app, "p6-ldar-cdk-projen-brand-image", { env: theEnv });
app.synth();
