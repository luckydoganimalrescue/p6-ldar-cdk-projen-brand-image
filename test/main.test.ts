import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { MyStack } from "../src/main";

test("Snapshot", () => {
  const theEnv = {
    account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
  };
  const app = new App();
  const stack = new MyStack(app, "test", { env: theEnv });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
