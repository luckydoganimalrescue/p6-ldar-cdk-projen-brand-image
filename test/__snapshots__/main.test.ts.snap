// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Certificate4E7ABB08": {
      "Properties": {
        "DomainName": "ldar.p6m7g8.net",
        "DomainValidationOptions": [
          {
            "DomainName": "ldar.p6m7g8.net",
            "ValidationDomain": "p6m7g8.net",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/Certificate",
          },
        ],
        "ValidationMethod": "EMAIL",
      },
      "Type": "AWS::CertificateManager::Certificate",
    },
    "DNFDC76583": {
      "Properties": {
        "DomainName": "ldar.p6m7g8.net",
        "DomainNameConfigurations": [
          {
            "CertificateArn": {
              "Ref": "Certificate4E7ABB08",
            },
            "EndpointType": "REGIONAL",
          },
        ],
      },
      "Type": "AWS::ApiGatewayV2::DomainName",
    },
    "HttpApiDefaultStage3EEB07D6": {
      "DependsOn": [
        "DNFDC76583",
      ],
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "HttpApiDefaultStagetestDNundefined4E93774E": {
      "DependsOn": [
        "DNFDC76583",
        "HttpApiDefaultStage3EEB07D6",
      ],
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "DomainName": {
          "Ref": "DNFDC76583",
        },
        "Stage": "$default",
      },
      "Type": "AWS::ApiGatewayV2::ApiMapping",
    },
    "HttpApiF5A9A8A7": {
      "Properties": {
        "CorsConfiguration": {
          "AllowHeaders": [
            "*",
          ],
          "AllowMethods": [
            "*",
          ],
          "AllowOrigins": [
            "*",
          ],
          "MaxAge": 864000,
        },
        "Description": "Pet Brander API",
        "Name": "HttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "HttpApiPOSTbrand280F256D": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /brand",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApiPOSTbrandbrandIntegrationFF0C65BA",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "HttpApiPOSTbrandbrandIntegrationFF0C65BA": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "brandE887937F",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "HttpApiPOSTbrandbrandIntegrationPermissionE5FBB5A5": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "brandE887937F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "HttpApiF5A9A8A7",
              },
              "/*/*/brand",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HttpApiPOSTpresign8A85FCBE": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "POST /presign",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "HttpApiPOSTpresignFuncIntegration8E43BEC9",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "HttpApiPOSTpresignFuncIntegration8E43BEC9": {
      "Properties": {
        "ApiId": {
          "Ref": "HttpApiF5A9A8A7",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "presignC1FD7464",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "HttpApiPOSTpresignFuncIntegrationPermission58F8471A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "presignC1FD7464",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "HttpApiF5A9A8A7",
              },
              "/*/*/presign",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "Identity2D60E2CC": {
      "Properties": {
        "EmailIdentity": "ldar-pet-brander@p6m7g8.com",
      },
      "Type": "AWS::SES::EmailIdentity",
    },
    "MyBucketF68F3FF0": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccessControl": "BucketOwnerFullControl",
        "CorsConfiguration": {
          "CorsRules": [
            {
              "AllowedHeaders": [
                "*",
              ],
              "AllowedMethods": [
                "GET",
                "PUT",
              ],
              "AllowedOrigins": [
                "*",
              ],
              "MaxAge": 3000,
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "IgnorePublicAcls": true,
        },
        "WebsiteConfiguration": {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "MyBucketPolicyE7FBAC7B": {
      "Properties": {
        "Bucket": {
          "Ref": "MyBucketF68F3FF0",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "MyBucketF68F3FF0",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "brandE887937F": {
      "DependsOn": [
        "brandServiceRoleDefaultPolicy520F6E98",
        "brandServiceRole1EE12E3C",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "726a21f99bce812a645f79d530bba544f19cd95c9c47400d7f70166158695e55.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BRAND_IMAGE_BUCKET": {
              "Ref": "MyBucketF68F3FF0",
            },
            "EMAIL_REGION": "us-east-1",
            "EMAIL_SENDER": "ldar-pet-brander@p6m7g8.com",
            "FIT": "fill",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "brandServiceRole1EE12E3C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 840,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "brandServiceRole1EE12E3C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "brandServiceRoleDefaultPolicy520F6E98": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "MyBucketF68F3FF0",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "MyBucketF68F3FF0",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ses:SendEmail",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "brandServiceRoleDefaultPolicy520F6E98",
        "Roles": [
          {
            "Ref": "brandServiceRole1EE12E3C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "presignC1FD7464": {
      "DependsOn": [
        "presignServiceRoleDefaultPolicy8E4B8563",
        "presignServiceRole70CB0056",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "8dd48f165e03802862b89d471e0962bbf0553c856bc3cde427c9af281c6f4af1.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "BUCKET_NAME": {
              "Ref": "MyBucketF68F3FF0",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "presignServiceRole70CB0056",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "presignServiceRole70CB0056": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "presignServiceRoleDefaultPolicy8E4B8563": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "MyBucketF68F3FF0",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "presignServiceRoleDefaultPolicy8E4B8563",
        "Roles": [
          {
            "Ref": "presignServiceRole70CB0056",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
